src/sentry/replays/lib/storage.py

```diff
@@ -0,0 +1,161 @@
+"""Replays storage blob driver module.
+
+Blob drivers are polymorphic on the service abstractions provided by Sentry.  Blob drivers are
+not polymorphic on service providers.  Any change of credentials will result in data loss unless
+proper steps are taken to migrate your data.
+"""
+import dataclasses
+import logging
+from abc import ABC, abstractmethod
+from datetime import datetime
+from io import BytesIO
+from typing import Optional, Union
+
+from django.conf import settings
+from django.db.utils import IntegrityError
+
+from sentry import options
+from sentry.models.file import File, get_storage
+from sentry.replays.cache import replay_cache
+from sentry.replays.models import ReplayRecordingSegment
+from sentry.utils import metrics
+
+logger = logging.getLogger()
+
+
+@dataclasses.dataclass
+class RecordingSegmentStorageMeta:
+    project_id: int
+    replay_id: str
+    segment_id: int
+    retention_days: int
+    date_added: Optional[datetime] = None
+    file_id: Optional[int] = None
+
+
+class Blob(ABC):
+    @abstractmethod
+    def delete(self, segment: RecordingSegmentStorageMeta) -> None:
+        """Remove a blob from remote storage."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self, segment: RecordingSegmentStorageMeta) -> bytes:
+        """Return blob from remote storage."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def set(self, segment: RecordingSegmentStorageMeta, value: bytes) -> None:
+        """Set blob in remote storage."""
+        raise NotImplementedError
+
+
+class FilestoreBlob(Blob):
+    """Filestore service driver blob manager."""
+
+    def delete(self, segment: RecordingSegmentStorageMeta) -> None:
+        file = File.objects.get(pk=segment.file_id)
+        file.delete()
+
+    @metrics.wraps("replays.lib.storage.FilestoreBlob.get")
+    def get(self, segment: RecordingSegmentStorageMeta) -> bytes:
+        file = File.objects.get(pk=segment.file_id)
+        return file.getfile().read()
+
+    @metrics.wraps("replays.lib.storage.FilestoreBlob.set")
+    def set(self, segment: RecordingSegmentStorageMeta, value: bytes) -> None:
+        file = File.objects.create(name=make_filename(segment), type="replay.recording")
+        file.putfile(BytesIO(value), blob_size=settings.SENTRY_ATTACHMENT_BLOB_SIZE)
+
+        try:
+            segment = ReplayRecordingSegment.objects.create(
+                project_id=segment.project_id,
+                replay_id=segment.replay_id,
+                segment_id=segment.segment_id,
+                file_id=file.id,
+                size=len(value),
+            )
+        except IntegrityError:
+            logger.warning(
+                "Recording-segment has already been processed.",
```

Comment:
we'll be losing this log with the new storage system which is fine. I wonder if the SDK folks could do a small investigation to see if they can replicate this happening, as it seems to happen more than duplicate kafka messages would cause. @billyvg @mydea

Reply:
seems to still happen a decent amount https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Alabels.name%3D~%22sentry%22%0A%22Recording-segment%20has%20already%20been%20processed.%22%0Alabels.name%3D%22sentry.replays%22;timeRange=PT3H;summaryFields=:true:32:beginning;cursorTimestamp=2023-02-28T21:21:29.109423150Z?project=internal-sentry&pli=1&rapt=AEjHL4OAUU7QN7hd54vapEcI538EIPYHa2nRP9IjsAEEKlSgkuBpznfLc52cKBN18RWG73ATAj_AwSE0ZxO_KR-zlNyEt5KsdA

Reply:
Could be old SDKs.  A good use case for our TBD internal metrics system.